server/schedulers/balance_leader.go

```diff
@@ -223,70 +209,92 @@ func (l *balanceLeaderScheduler) Schedule(cluster schedule.Cluster) []*operator.
-		desc:          "target",
-		actionDesc:    "transfer-in",
 		compareOption: lessOption,
-		setPlanStore: func(bp *balancePlan, si *core.StoreInfo) {
-			plan.source, plan.target = nil, si
-		},
-		opCreator:     l.transferLeaderIn,
-		schedulerName: l.GetName(),
-		plan:          plan,
 	}
 	targetCandidate.initSort()
 	usedRegions := make(map[uint64]struct{})
-	candidates := []*candidateStores{sourceCandidate, targetCandidate}
 
 	result := make([]*operator.Operator, 0, l.conf.Batch)
 	for hasStore := true; hasStore; {
 		hasStore = false
-		for _, queue := range candidates {
-			if queue.hasStore() {
-				hasStore = true
-				store := queue.getStore()
-				retryLimit := l.retryQuota.GetLimit(store)
-				log.Debug("store leader score", zap.String("scheduler", l.GetName()), zap.Uint64(queue.desc, store.GetID()))
-				l.counter.WithLabelValues("high-score", strconv.FormatUint(store.GetID(), 10)).Inc()
-				op := queue.generateScheduleOperator(retryLimit, usedRegions)
-				if op != nil {
-					l.retryQuota.ResetLimit(store)
-					op.Counters = append(op.Counters, l.counter.WithLabelValues(queue.actionDesc, plan.SourceMetricLabel()))
-					result = append(result, op)
-					if len(result) >= l.conf.Batch {
-						return result
-					}
-					// the following is used to make influence
-					usedRegions[op.RegionID()] = struct{}{}
-					schedule.AddOpInfluence(op, plan.opInfluence, cluster)
-					for _, queue := range candidates {
-						queue.reSort(plan.source)
-						queue.reSort(plan.target)
-					}
-				} else {
-					l.Attenuate(store)
-					log.Debug("no operator created for selected stores", zap.String("scheduler", l.GetName()), zap.Uint64(queue.desc, queue.getStore().GetID()))
-					queue.next()
+		// first choose source
+		if sourceCandidate.hasStore() {
+			hasStore = true
+			op := createTransferLeaderOperator(sourceCandidate, transferOut, l, plan, usedRegions)
+			if op != nil {
+				result = append(result, op)
+				if len(result) >= l.conf.Batch {
+					return result
+				}
+				makeInfluence(op, plan, usedRegions, sourceCandidate, targetCandidate)
+			}
+		}
+		// next choose target
+		if targetCandidate.hasStore() {
+			hasStore = true
+			op := createTransferLeaderOperator(targetCandidate, transferIn, l, plan, usedRegions)
+			if op != nil {
+				result = append(result, op)
+				if len(result) >= l.conf.Batch {
+					return result
 				}
+				makeInfluence(op, plan, usedRegions, sourceCandidate, targetCandidate)
 			}
 		}
 	}
 	l.retryQuota.GC(append(sourceCandidate.stores, targetCandidate.stores...))
 	return result
 }
 
+func createTransferLeaderOperator(cs *candidateStores, dir string, l *balanceLeaderScheduler,
+	plan *balancePlan, usedRegions map[uint64]struct{}) *operator.Operator {
+	store := cs.getStore()
+	retryLimit := l.retryQuota.GetLimit(store)
+	var creator func(*balancePlan) *operator.Operator
+	log.Debug("store leader score", zap.String("scheduler", l.GetName()), zap.Uint64(dir, store.GetID()))
```

Comment:
no score field?

Reply:
This statement didn't have it before, but it's better to add it.